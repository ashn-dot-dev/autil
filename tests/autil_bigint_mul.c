#define AUTIL_IMPLEMENTATION
#include "../autil.h"

struct autil_bigint* RES = NULL;
struct autil_bigint* LHS = NULL;
struct autil_bigint* RHS = NULL;

static void
test_autil_bigint_mul__0_mul_x(void)
{
    puts(__func__);

    RES = autil_bigint_new(AUTIL_BIGINT_ZERO);
    autil_bigint_mul(RES, RES, AUTIL_BIGINT_POS_ONE);
    autil_bigint_dump(RES);
    autil_bigint_del(RES);

    RES = autil_bigint_new(AUTIL_BIGINT_ZERO);
    autil_bigint_mul(RES, RES, AUTIL_BIGINT_NEG_ONE);
    autil_bigint_dump(RES);
    autil_bigint_del(RES);
}

static void
test_autil_bigint_mul__x_mul_0(void)
{
    puts(__func__);

    RES = autil_bigint_new(AUTIL_BIGINT_ZERO);
    autil_bigint_mul(RES, AUTIL_BIGINT_POS_ONE, RES);
    autil_bigint_dump(RES);
    autil_bigint_del(RES);

    RES = autil_bigint_new(AUTIL_BIGINT_ZERO);
    autil_bigint_mul(RES, AUTIL_BIGINT_NEG_ONE, RES);
    autil_bigint_dump(RES);
    autil_bigint_del(RES);
}

static void
test_autil_bigint_mul__pos_mul_pos(void)
{
    puts(__func__);

    RES = autil_bigint_new(AUTIL_BIGINT_ZERO);
    autil_bigint_mul(RES, AUTIL_BIGINT_POS_ONE, AUTIL_BIGINT_POS_ONE);
    autil_bigint_dump(RES);
    autil_bigint_del(RES);

    RES = autil_bigint_new(AUTIL_BIGINT_ZERO);
    LHS = autil_bigint_new_cstr("+0xffee");
    RHS = autil_bigint_new_cstr("+0xffee");
    autil_bigint_mul(RES, LHS, RHS);
    autil_bigint_dump(RES);
    autil_bigint_del(RES);
    autil_bigint_del(LHS);
    autil_bigint_del(RHS);
}

static void
test_autil_bigint_mul__neg_mul_neg(void)
{
    puts(__func__);

    RES = autil_bigint_new(AUTIL_BIGINT_ZERO);
    autil_bigint_mul(RES, AUTIL_BIGINT_NEG_ONE, AUTIL_BIGINT_NEG_ONE);
    autil_bigint_dump(RES);
    autil_bigint_del(RES);

    RES = autil_bigint_new(AUTIL_BIGINT_ZERO);
    LHS = autil_bigint_new_cstr("-0xffee");
    RHS = autil_bigint_new_cstr("-0xffee");
    autil_bigint_mul(RES, LHS, RHS);
    autil_bigint_dump(RES);
    autil_bigint_del(RES);
    autil_bigint_del(LHS);
    autil_bigint_del(RHS);
}

static void
test_autil_bigint_mul__pos_mul_neg(void)
{
    puts(__func__);

    RES = autil_bigint_new(AUTIL_BIGINT_ZERO);
    autil_bigint_mul(RES, AUTIL_BIGINT_POS_ONE, AUTIL_BIGINT_NEG_ONE);
    autil_bigint_dump(RES);
    autil_bigint_del(RES);

    RES = autil_bigint_new(AUTIL_BIGINT_ZERO);
    LHS = autil_bigint_new_cstr("+0xffee");
    RHS = autil_bigint_new_cstr("-0xffee");
    autil_bigint_mul(RES, LHS, RHS);
    autil_bigint_dump(RES);
    autil_bigint_del(RES);
    autil_bigint_del(LHS);
    autil_bigint_del(RHS);
}

static void
test_autil_bigint_mul__neg_mul_pos(void)
{
    puts(__func__);

    RES = autil_bigint_new(AUTIL_BIGINT_ZERO);
    autil_bigint_mul(RES, AUTIL_BIGINT_NEG_ONE, AUTIL_BIGINT_POS_ONE);
    autil_bigint_dump(RES);
    autil_bigint_del(RES);

    RES = autil_bigint_new(AUTIL_BIGINT_ZERO);
    LHS = autil_bigint_new_cstr("-0xffee");
    RHS = autil_bigint_new_cstr("+0xffee");
    autil_bigint_mul(RES, LHS, RHS);
    autil_bigint_dump(RES);
    autil_bigint_del(RES);
    autil_bigint_del(LHS);
    autil_bigint_del(RHS);
}

int
main(void)
{
    test_autil_bigint_mul__0_mul_x();
    test_autil_bigint_mul__x_mul_0();

    test_autil_bigint_mul__pos_mul_pos();
    test_autil_bigint_mul__neg_mul_neg();
    test_autil_bigint_mul__pos_mul_neg();
    test_autil_bigint_mul__neg_mul_pos();

    return EXIT_SUCCESS;
}
