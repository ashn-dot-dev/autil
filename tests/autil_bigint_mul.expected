test_autil_bigint_mul__0_mul_x
SIGN: '0', COUNT: 0, LIMBS: []
SIGN: '0', COUNT: 0, LIMBS: []
test_autil_bigint_mul__x_mul_0
SIGN: '0', COUNT: 0, LIMBS: []
SIGN: '0', COUNT: 0, LIMBS: []
test_autil_bigint_mul__pos_mul_pos
SIGN: '+', COUNT: 1, LIMBS: [0x01]
SIGN: '+', COUNT: 4, LIMBS: [0x44, 0x01, 0xDC, 0xFF]
test_autil_bigint_mul__neg_mul_neg
SIGN: '+', COUNT: 1, LIMBS: [0x01]
SIGN: '+', COUNT: 4, LIMBS: [0x44, 0x01, 0xDC, 0xFF]
test_autil_bigint_mul__pos_mul_neg
SIGN: '-', COUNT: 1, LIMBS: [0x01]
SIGN: '-', COUNT: 4, LIMBS: [0x44, 0x01, 0xDC, 0xFF]
test_autil_bigint_mul__neg_mul_pos
SIGN: '-', COUNT: 1, LIMBS: [0x01]
SIGN: '-', COUNT: 4, LIMBS: [0x44, 0x01, 0xDC, 0xFF]
